server/channels/api4/channel.go

```diff
@@ -425,9 +425,14 @@ func restoreChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 }
 
 func createDirectChannel(c *Context, w http.ResponseWriter, r *http.Request) {
-	userIds := model.ArrayFromJSON(r.Body)
+	userIds := model.ArrayFromJSONNonSort(r.Body, *c.App.Config().ServiceSettings.MaximumPayloadSize)
```

Comment:
I wonder if we might want to force a limit reader for all API handlers by default by modifying `r.Body` before we even get here. In `ServeHTTP`, we could do something like
```go
r.Body = io.LimitReader(r.Body, maximumPayloadSize)
```
to have this apply anywhere. Maybe it could hold a reference to the original `r.Body` somehow so that the file upload APIs could get around this, but it would mean that we only need to make one change there and it'd apply automatically to all new APIs down the line

Reply:
@hmhealey - Thanks for suggestion. Updating in `ServeHTTP` was discussed in [this document](https://docs.google.com/document/d/1v8nhBM10ZMsfzO8yovTCu3fcZiYt5prQLBSpYPnmjUg/edit). I'm willing to revisit, as we want to get this correct. 

We should also discuss the breaking change aspect. Is that ok for a security issue?

Reply:
Thanks for the link. I think doing it from `ServeHTTP` would make it a bit less likely that we accidentally re-introduce this issue down the line, but I don't think makes a huge difference either way.

As for the breaking change, I'm 0/5 on that since I don't know much about the Go ecosystem there. I just saw that something in the public package was changing, so I called it out. If we do care about it, we could just deprecate the current version and give your new ones a new name

Reply:
@hmhealey - We discussed this. There are pros and cons to both approaches. Separating the file upload limit with the input body limit will require a separate middleware and a property to indicate which limit should be respected. We wanted to have a layered approach here - where at a high level, the file upload limit is respected. But different handlers are free to choose their own limit depending on the application logic.

To not re-introduce this again, we would be relying on vet checks to ensure people don't forget. 

There's more discussion on the google docs.